subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java

```diff
@@ -122,78 +106,48 @@ public void triggerWatchProbe(String path) {
+        watchedFiles = updateWatchedHierarchies(root);
 
         // Probe every hierarchy that is watched, even ones nested inside others
+        ImmutableSet<File> oldProbedHierarchies = probedHierarchies;
         probedHierarchies = watchableHierarchies.stream()
             .filter(watchedFiles::contains)
             .collect(ImmutableSet.toImmutableSet());
-
-        if (oldWatchedHierarchies.equals(watchedHierarchies) && oldProbedHierarchies.equals(probedHierarchies)) {
-            // Nothing changed
+        if (oldProbedHierarchies.equals(probedHierarchies)) {
             return;
         }
 
-        if (watchedHierarchies.isEmpty()) {
-            LOGGER.info("Not watching anything anymore");
-        }
-        List<File> hierarchiesToStopWatching = oldWatchedHierarchies.stream()
-            .filter(oldWatchedHierarchy -> !watchedHierarchies.contains(oldWatchedHierarchy))
-            .collect(Collectors.toCollection(() -> new ArrayList<>(oldWatchedHierarchies.size())));
-        List<File> hierarchiesToStartWatching = watchedHierarchies.stream()
-            .filter(newWatchedHierarchy -> !oldWatchedHierarchies.contains(newWatchedHierarchy))
-            .collect(Collectors.toCollection(() -> new ArrayList<>(watchedHierarchies.size())));
-
         oldProbedHierarchies.stream()
             .filter(oldProbedHierarchy -> !probedHierarchies.contains(oldProbedHierarchy))
-            .forEach(probedHierarchy -> {
-                stopWatchingProbeForHierarchy(probedHierarchy);
-                probeRegistry.disarmWatchProbe(probedHierarchy);
-            });
-
-        if (!hierarchiesToStopWatching.isEmpty()) {
-            stopWatchingHierarchies(hierarchiesToStopWatching);
-        }
-        if (!hierarchiesToStartWatching.isEmpty()) {
-            hierarchiesToStartWatching.forEach(locationToWatchValidator::validateLocationToWatch);
-            startWatchingHierarchies(hierarchiesToStartWatching);
-        }
+            .forEach(this::disarmWatchProbeForHierarchy);
 
         probedHierarchies.stream()
             .filter(newProbedHierarchy -> !oldProbedHierarchies.contains(newProbedHierarchy))
-            .forEach(probedHierarchy -> {
-                startWatchingProbeForHierarchy(probedHierarchy);
-                probeRegistry.armWatchProbe(probedHierarchy);
-            });
-
-        LOGGER.info("Watching {} directory hierarchies to track changes", watchedHierarchies.size());
+            .forEach(this::armWatchProbeForHierarchy);
     }
 
-    protected abstract void startWatchingHierarchies(Collection<File> hierarchiesToWatch);
-
-    protected abstract void startWatchingProbeForHierarchy(File hierarchyToWatch);
-
-    protected abstract void stopWatchingHierarchies(Collection<File> hierarchiesToWatch);
-
-    protected abstract void stopWatchingProbeForHierarchy(File hierarchyToWatch);
+    protected FileHierarchySet updateWatchedHierarchies(SnapshotHierarchy root) {
```

Comment:
Should this also have `@OverridingMethodsMustInvokeSuper`?

Reply:
Yes, added.